// frontend/src/pages/NewArrivals.jsx
import React from "react";
import { motion } from "framer-motion";
import Navbar from "../components/Navbar";
import heroImg from "../assets/hero.jpg";

// --- Data ---
const arrivals = [
  {
    id: 1,
    name: "Dew Barrier Essence",
    price: "₹1,299",
    tag: "Just Dropped",
    note: "Rice peptides + ceramide support",
  },
  {
    id: 2,
    name: "Cloud Clean Gel",
    price: "₹899",
    tag: "New",
    note: "Low-foam cleanser for daily reset",
  },
  {
    id: 3,
    name: "Peptide Lift Mist",
    price: "₹1,099",
    tag: "Trending",
    note: "Fine hydration veil with bounce",
  },
  {
    id: 4,
    name: "Luminous Night Oil",
    price: "₹1,699",
    tag: "Limited",
    note: "Squalane-rich overnight nourishment",
  },
  {
    id: 5,
    name: "Mineral Veil SPF 45",
    price: "₹999",
    tag: "New",
    note: "Weightless UV shield, no cast",
  },
  {
    id: 6,
    name: "Calm Rescue Cream",
    price: "₹1,249",
    tag: "Just Dropped",
    note: "Barrier-first comfort for redness",
  },
];

// --- Animation Variants ---
const fadeUp = {
  hidden: { opacity: 0, y: 40 },
  show: { opacity: 1, y: 0, transition: { duration: 0.8, ease: [0.16, 1, 0.3, 1] } },
};

const staggerContainer = {
  hidden: { opacity: 0 },
  show: {
    opacity: 1,
    transition: { staggerChildren: 0.12 },
  },
};

// --- Sub-components ---
const ProductCard = ({ item }) => {
  return (
    <motion.article
      variants={fadeUp}
      className="group cursor-pointer flex flex-col items-center text-center"
    >
      {/* Luminous Image Container */}
      <div className="relative w-full aspect-[4/5] overflow-hidden rounded-[2.5rem] bg-gradient-to-b from-slate-50 to-slate-100 mb-8 transition-all duration-700 ease-out group-hover:shadow-[0_30px_60px_-15px_rgba(0,0,0,0.08)] group-hover:-translate-y-2">
        <img
          src={heroImg}
          alt={item.name}
          className="absolute inset-0 h-full w-full object-cover transition-transform duration-1000 ease-[0.16,1,0.3,1] group-hover:scale-105 opacity-90 mix-blend-multiply"
        />
        
        {/* Soft Glass Overlay */}
        <div className="absolute inset-0 bg-white/0 transition-colors duration-500 group-hover:bg-white/10 backdrop-blur-[1px] opacity-0 group-hover:opacity-100" />
        
        {/* Floating Minimal Tag */}
        <div className="absolute top-6 right-6">
          <span className="bg-white/80 backdrop-blur-md text-slate-800 text-[9px] font-semibold tracking-[0.2em] uppercase px-4 py-2 rounded-full shadow-sm">
            {item.tag}
          </span>
        </div>

        {/* Ethereal Add Button */}
        <div className="absolute inset-x-6 bottom-6 translate-y-8 opacity-0 group-hover:translate-y-0 group-hover:opacity-100 transition-all duration-500 ease-[0.16,1,0.3,1]">
          <button className="w-full py-4 bg-white/90 backdrop-blur-xl text-slate-900 text-xs font-semibold tracking-[0.15em] uppercase rounded-2xl hover:bg-white transition-colors shadow-[0_8px_30px_rgb(0,0,0,0.12)]">
            Add to Cart — {item.price}
          </button>
        </div>
      </div>

      {/* Airy Content */}
      <div className="flex flex-col px-4 w-full">
        <h3 className="text-xl font-light tracking-tight text-slate-900">
          {item.name}
        </h3>
        <p className="mt-2 text-sm text-slate-500 font-light leading-relaxed">
          {item.note}
        </p>
      </div>
    </motion.article>
  );
};

// --- Main Page Component ---
export default function NewArrivals() {
  return (
    <div className="min-h-screen bg-white text-slate-900 font-sans selection:bg-rose-100 selection:text-slate-900">
      <Navbar />

      {/* LUMINOUS HERO SECTION */}
      <section className="relative w-full h-[85svh] flex items-center justify-center overflow-hidden pt-20">
        {/* Soft, glowing orb backgrounds instead of harsh images */}
        <div className="absolute top-[-10%] left-[-10%] w-[50vw] h-[50vw] rounded-full bg-rose-50 blur-[120px] opacity-70" />
        <div className="absolute bottom-[-10%] right-[-10%] w-[50vw] h-[50vw] rounded-full bg-teal-50 blur-[120px] opacity-70" />
        <div className="absolute top-[20%] left-[40%] w-[40vw] h-[40vw] rounded-full bg-amber-50 blur-[100px] opacity-60" />

        <motion.div
          variants={staggerContainer}
          initial="hidden"
          animate="show"
          className="relative z-10 max-w-4xl flex flex-col items-center justify-center text-center px-6"
        >
          <motion.div variants={fadeUp} className="mb-8">
            <span className="text-[10px] sm:text-xs tracking-[0.3em] uppercase text-slate-400 font-medium border-b border-slate-200 pb-2">
              Spring / Summer Collection
            </span>
          </motion.div>

          <motion.h1 variants={fadeUp} className="text-5xl md:text-7xl lg:text-8xl font-light tracking-tighter text-slate-900 leading-[1.1]">
            Meet the <span className="italic font-serif text-slate-400">Future</span> of <br/>Your Skin.
          </motion.h1>
          
          <motion.p variants={fadeUp} className="mt-8 max-w-xl text-base leading-relaxed text-slate-500 md:text-lg font-light">
            Weightless textures. Uncompromising clinical efficacy. Discover the new arrivals formulated to let your natural barrier breathe and thrive.
          </motion.p>

          <motion.button variants={fadeUp} className="mt-12 px-8 py-4 rounded-full bg-slate-900 text-white text-xs font-semibold tracking-[0.2em] uppercase hover:bg-slate-800 hover:-translate-y-1 transition-all duration-300 shadow-xl shadow-slate-900/20">
            Explore the Drop
          </motion.button>
        </motion.div>
      </section>

      {/* AIRY GRID SECTION */}
      <section className="mx-auto max-w-7xl px-6 py-24 md:px-12 lg:px-24">
        {/* Minimal Section Header */}
        <motion.div 
          initial={{ opacity: 0, y: 20 }}
          whileInView={{ opacity: 1, y: 0 }}
          viewport={{ once: true }}
          transition={{ duration: 0.8 }}
          className="mb-20 text-center flex flex-col items-center"
        >
          <h2 className="text-3xl md:text-4xl font-light tracking-tight text-slate-900">
            Freshly <span className="font-serif italic text-slate-400">Formulated</span>
          </h2>
          <div className="h-px w-12 bg-slate-200 mt-6" />
        </motion.div>

        {/* Animated Grid with Staggered Layout feel */}
        <motion.div 
          variants={staggerContainer}
          initial="hidden"
          whileInView="show"
          viewport={{ once: true, margin: "-100px" }}
          className="grid gap-x-8 gap-y-20 sm:grid-cols-2 lg:grid-cols-3"
        >
          {arrivals.map((item) => (
            <ProductCard key={item.id} item={item} />
          ))}
        </motion.div>
      </section>

      {/* SOFT & BRIGHT FOOTER */}
      <footer className="bg-slate-50 text-slate-900 pt-24 pb-12 px-6 md:px-12 lg:px-24 mt-12 border-t border-slate-100">
        <div className="grid grid-cols-1 lg:grid-cols-12 gap-16 mb-20">
          
          {/* Newsletter */}
          <div className="lg:col-span-6 pr-0 lg:pr-12">
            <h2 className="text-3xl font-light mb-4 tracking-tight">The <span className="font-serif italic text-slate-400">Sanctuary</span> List</h2>
            <p className="text-slate-500 font-light mb-8 max-w-md">Early access to new drops, clinical deep-dives, and 10% off your first ritual.</p>
            <div className="flex relative max-w-md">
              <input 
                type="email" 
                placeholder="Enter your email" 
                className="w-full bg-transparent border-b border-slate-300 text-slate-900 placeholder-slate-400 px-0 py-4 focus:outline-none focus:border-slate-900 transition-colors rounded-none"
              />
              <button className="absolute right-0 top-1/2 -translate-y-1/2 text-xs font-semibold uppercase tracking-widest text-slate-900 hover:text-slate-500 transition-colors">
                Subscribe
              </button>
            </div>
          </div>

          {/* Links Grid */}
          <div className="lg:col-span-6 grid grid-cols-2 md:grid-cols-3 gap-8 text-sm font-light text-slate-500">
            <div className="flex flex-col gap-4">
              <h4 className="text-slate-900 font-medium tracking-widest uppercase text-xs mb-2">Shop</h4>
              <a href="#" className="hover:text-slate-900 transition-colors">All Products</a>
              <a href="#" className="hover:text-slate-900 transition-colors">Best Sellers</a>
              <a href="#" className="hover:text-slate-900 transition-colors">New Arrivals</a>
              <a href="#" className="hover:text-slate-900 transition-colors">Routines</a>
            </div>
            <div className="flex flex-col gap-4">
              <h4 className="text-slate-900 font-medium tracking-widest uppercase text-xs mb-2">About</h4>
              <a href="#" className="hover:text-slate-900 transition-colors">Our Story</a>
              <a href="#" className="hover:text-slate-900 transition-colors">Ingredients</a>
              <a href="#" className="hover:text-slate-900 transition-colors">Clinical Trials</a>
            </div>
            <div className="flex flex-col gap-4">
              <h4 className="text-slate-900 font-medium tracking-widest uppercase text-xs mb-2">Support</h4>
              <a href="#" className="hover:text-slate-900 transition-colors">FAQ</a>
              <a href="#" className="hover:text-slate-900 transition-colors">Shipping & Returns</a>
              <a href="#" className="hover:text-slate-900 transition-colors">Contact</a>
            </div>
          </div>
        </div>

        <div className="flex flex-col md:flex-row justify-between items-center pt-8 border-t border-slate-200 text-xs text-slate-400 font-light">
          <p>© 2026 Clinical Sanctuary. All rights reserved.</p>
          <div className="flex gap-6 mt-4 md:mt-0">
            <a href="#" className="hover:text-slate-900 transition-colors">Privacy</a>
            <a href="#" className="hover:text-slate-900 transition-colors">Terms</a>
            <a href="#" className="hover:text-slate-900 transition-colors">Instagram</a>
          </div>
        </div>
      </footer>
    </div>
  );
}
